Subject: [Bug 429490] sles10 raid0 can not be started                                                                                                                  
From: bugzilla_noreply@novell.com                                                                                                                                      

### Comment #4 from Neil Brown <nfbrown@novell.com>  2008-10-08 17:34:49 MDT ---
The md code was written with the expectation that the chunk size would
always be at least one page.  This may not always be a strict requirement,
but I guess that test protects against untested code.

For raid4/5/6, the stripe-cache is stored as pages so a chunk size
less than one page would certainly cause problems with the current code.

For raid0, I think a smaller chunk size should work.
The filesystem is always allowed to send down one-page requests, and
if the chunk size is less than that page size, these requests will have
to be split up, possibly multiple times.  However I think the code will
get this right - it has only actually be used for page-sized requests
that are not page-aligned.

You could try changing the test to

  if (chunk_size < PAGE_SIZE && mddev->level != 0) {

and see how that works.

Signed-off-by: Olaf Hering <olh@suse.de>

---
 drivers/md/md.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -3543,9 +3543,14 @@ static int do_md_run(mddev_t * mddev)
 			return -EINVAL;
 		}
 		if (chunk_size < PAGE_SIZE) {
+			if (mddev->level) {
 			printk(KERN_ERR "too small chunk_size: %d < %ld\n",
 				chunk_size, PAGE_SIZE);
 			return -EINVAL;
+			} else {
+				printk(KERN_ERR "too small chunk_size: %d < %ld, but continuing anyway on raid0. Good luck!\n",
+					chunk_size, PAGE_SIZE);
+			}
 		}
 
 		/* devices must have minimum size of one chunk */
